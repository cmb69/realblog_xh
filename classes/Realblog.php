<?php

/**
 * Copyright 2006-2010 Jan Kanters
 * Copyright 2010-2014 Gert Ebersbach
 * Copyright 2014-2017 Christoph M. Becker
 *
 * This file is part of Realblog_XH.
 *
 * Realblog_XH is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * Realblog_XH is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Realblog_XH.  If not, see <http://www.gnu.org/licenses/>.
 */

namespace Realblog;

use stdClass;
use ReflectionClass;

class Realblog
{
    const VERSION = '3.0beta9';

    /**
     * @return void
     */
    public static function init()
    {
        global $su, $plugin_cf, $plugin_tx;

        self::registerCommands();
        if ($plugin_cf['realblog']['auto_publish']) {
            self::autoPublish();
        }
        if ($plugin_cf['realblog']['auto_archive']) {
            self::autoArchive();
        }
        if ($plugin_cf['realblog']['rss_enabled']) {
            self::emitAlternateRSSLink();
            $rssFeedRequested = filter_input(
                INPUT_GET,
                'realblog_feed',
                FILTER_VALIDATE_REGEXP,
                array('options' => array('regexp' => '/^rss$/'))
            );
            if ($rssFeedRequested) {
                $controller = new FeedController($plugin_cf['realblog'], $plugin_tx['realblog'], new Finder());
                echo $controller->defaultAction();
                exit;
            }
        }
        /** @psalm-suppress UndefinedConstant */
        if (XH_ADM) {
            self::registerPluginMenu();
            if (XH_wantsPluginAdministration('realblog') || $su === 'realblog') {
                self::handleAdministration();
            }
        }
    }

    /**
     * @return void
     */
    private static function registerCommands()
    {
        $class = new ReflectionClass(Realblog::class);
        $commands = array();
        foreach ($class->getMethods() as $method) {
            $methodName = $method->getName();
            if (preg_match('/.*(?=Command$)/', $methodName, $m)) {
                $commands[$m[0]] = $method->getParameters();
            }
        }
        foreach ($commands as $name => $params) {
            $paramList = $argList = array();
            foreach ($params as $param) {
                $string = '$' . $param->getName();
                $argList[] = $string;
                if ($param->isOptional()) {
                    $string .= '=' . var_export($param->getDefaultValue(), true);
                }
                $paramList[] = $string;
            }
            $paramString = implode(',', $paramList);
            $argString = implode(',', $argList);
            eval("function Realblog_$name($paramString) {return \\Realblog\\Realblog::{$name}Command($argString);}");
        }
    }

    /**
     * @return void
     */
    private static function registerPluginMenu()
    {
        global $sn, $plugin_tx;

        XH_registerStandardPluginMenuItems(true);
        XH_registerPluginMenuItem(
            'realblog',
            $plugin_tx['realblog']['exchange_heading'],
            "$sn?realblog&admin=data_exchange"
        );
    }

    /**
     * @return void
     */
    private static function handleAdministration()
    {
        global $sn, $admin, $o, $plugin_tx;

        $o .= print_plugin_admin('on');
        pluginMenu('ROW');
        pluginMenu('TAB', "$sn?realblog&admin=data_exchange", '', $plugin_tx['realblog']['exchange_heading']);
        $o .= pluginMenu('SHOW');
        switch ($admin) {
            case '':
                $o .= (new InfoController(new View()))->defaultAction();
                break;
            case 'plugin_main':
                self::routeTo(MainAdminController::class);
                break;
            case 'data_exchange':
                self::routeTo(DataExchangeController::class);
                break;
            default:
                $o .= plugin_admin_common();
        }
    }

    /**
     * @param class-string $controllerClassName
     * @return void
     */
    private static function routeTo($controllerClassName)
    {
        global $o, $action, $plugin_cf, $plugin_tx, $_XH_csrfProtection;

        $methodName = lcfirst(implode('', array_map('ucfirst', explode('_', $action)))) . 'Action';
        $controller = new $controllerClassName(
            $plugin_cf['realblog'],
            $plugin_tx['realblog'],
            new Finder(),
            $_XH_csrfProtection,
            new View()
        );
        $class = new ReflectionClass($controller);
        if ($class->hasMethod($methodName)
            && ($method = $class->getMethod($methodName))
            && $method->isPublic()
        ) {
            $o .= $method->invoke($controller);
        } else {
            $o .= $controller->defaultAction();
        }
    }

    /**
     * @return void
     * @global string $hjs
     */
    private static function emitAlternateRSSLink()
    {
        global $hjs;

        $hjs .= '<link rel="alternate" type="application/rss+xml"'
            . ' href="./?realblog_feed=rss">';
    }

    /**
     * @return void
     */
    private static function autoPublish()
    {
        DB::autoChangeStatus('publishing_date', 1);
    }

    /**
     * @return void
     */
    private static function autoArchive()
    {
        DB::autoChangeStatus('archiving_date', 2);
    }

    /**
     * @param bool $showSearch
     * @param string $category
     * @return string
     */
    public static function blogCommand($showSearch = false, $category = 'all')
    {
        global $plugin_cf, $plugin_tx;

        $controller = new BlogController(
            $plugin_cf['realblog'],
            $plugin_tx['realblog'],
            $showSearch,
            new Finder(),
            new View(),
            $category
        );
        if (filter_has_var(INPUT_GET, 'realblog_id')) {
            return (string) $controller->showArticleAction(filter_input(
                INPUT_GET,
                'realblog_id',
                FILTER_VALIDATE_INT,
                array('options' => array('min_range' => 1))
            ));
        } else {
            return $controller->defaultAction();
        }
    }

    /**
     * @param bool $showSearch
     * @return string
     */
    public static function archiveCommand($showSearch = false)
    {
        global $plugin_cf, $plugin_tx;

        $controller = new ArchiveController(
            $plugin_cf['realblog'],
            $plugin_tx['realblog'],
            $showSearch,
            new Finder(),
            new View()
        );
        if (filter_has_var(INPUT_GET, 'realblog_id')) {
            return (string) $controller->showArticleAction(filter_input(
                INPUT_GET,
                'realblog_id',
                FILTER_VALIDATE_INT,
                array('options' => array('min_range' => 1))
            ));
        } else {
            return $controller->defaultAction();
        }
    }

    /**
     * @param string $pageUrl
     * @param bool $showTeaser
     * @return string
     */
    public static function linkCommand($pageUrl, $showTeaser = false)
    {
        global $plugin_cf, $plugin_tx;

        $controller = new LinkController(
            $plugin_cf['realblog'],
            $plugin_tx['realblog'],
            $pageUrl,
            $showTeaser,
            new Finder(),
            new View()
        );
        return $controller->defaultAction();
    }

    /**
     * @param string $pageUrl
     * @return string
     */
    public static function mostPopularCommand($pageUrl)
    {
        global $plugin_cf, $plugin_tx;

        $controller = new MostPopularController(
            $plugin_cf['realblog'],
            $plugin_tx['realblog'],
            $pageUrl,
            new Finder(),
            new View()
        );
        return $controller->defaultAction();
    }

    /**
     * @param string $target
     * @return string
     */
    public static function feedLinkCommand($target = '_self')
    {
        global $plugin_cf, $plugin_tx;

        $controller = new FeedLinkController($plugin_cf['realblog'], $plugin_tx['realblog']);
        return $controller->defaultAction($target);
    }

    /**
     * @return int
     */
    public static function getPage()
    {
        global $edit;

        /** @psalm-suppress UndefinedConstant */
        if (XH_ADM && $edit) {
            if (filter_has_var(INPUT_GET, 'realblog_page')) {
                $page = filter_input(
                    INPUT_GET,
                    'realblog_page',
                    FILTER_VALIDATE_INT,
                    array('options' => array('min_range' => 1))
                );
                $_COOKIE['realblog_page'] = $page;
                setcookie('realblog_page', $page, 0, CMSIMPLE_ROOT);
            } else {
                $page = filter_input(
                    INPUT_COOKIE,
                    'realblog_page',
                    FILTER_VALIDATE_INT,
                    array('options' => array('min_range' => 1, 'default' => 1))
                );
            }
        } else {
            $page = filter_input(
                INPUT_GET,
                'realblog_page',
                FILTER_VALIDATE_INT,
                array('options' => array('min_range' => 1, 'default' => 1))
            );
        }
        return $page;
    }

    /**
     * @param int $num
     * @return bool
     */
    public static function getFilter($num)
    {
        $varname = "realblog_filter$num";
        if (filter_has_var(INPUT_GET, $varname)) {
            $filter = filter_input(INPUT_GET, $varname, FILTER_VALIDATE_BOOLEAN);
            $_COOKIE[$varname] = $filter ? 'on' : '';
            setcookie($varname, $filter ? 'on' : '', 0, CMSIMPLE_ROOT);
        } else {
            $filter = filter_input(INPUT_COOKIE, $varname, FILTER_VALIDATE_BOOLEAN);
        }
        return $filter;
    }

    /**
     * @param string $pageUrl
     * @param array  $params
     * @return string
     */
    public static function url($pageUrl, $params = array())
    {
        global $sn;

        $replacePairs = array(
            //'realblog_id' => 'id',
            //'realblog_page' => 'page'
        );
        $url = $sn . '?' . $pageUrl;
        ksort($params);
        foreach ($params as $name => $value) {
            if (!($name == 'realblog_page' && $value == 1
                || $name == 'realblog_year' && $value == date('Y')
                || $name == 'realblog_search' && !$value)
            ) {
                $url .= '&' . strtr($name, $replacePairs) . '=' . $value;
            }
        }
        return $url;
    }
}
