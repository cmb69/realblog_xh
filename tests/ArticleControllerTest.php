<?php

/**
 * Copyright 2023 Christoph M. Becker
 *
 * This file is part of Realblog_XH.
 *
 * Realblog_XH is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * Realblog_XH is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Realblog_XH.  If not, see <http://www.gnu.org/licenses/>.
 */

namespace Realblog;

use ApprovalTests\Approvals;
use PHPUnit\Framework\MockObject\MockObject;
use PHPUnit\Framework\MockObject\Stub;
use PHPUnit\Framework\TestCase;
use Plib\FakeRequest;
use Plib\View;
use Realblog\Infra\DB;
use Realblog\Infra\FakePages;
use Realblog\Infra\Finder;
use Realblog\Value\Article;
use Realblog\Value\FullArticle;

class ArticleControllerTest extends TestCase
{
    /** @var array<string,string> */
    private $conf;

    /** @var Finder&Stub */
    private $finder;

    /** @var DB&MockObject */
    private $db;

    /** @var FakePages */
    private $pages;

    /** @var View */
    private $view;

    public function setUp(): void
    {
        $this->conf = XH_includeVar("./config/config.php", "plugin_cf")["realblog"];
        $this->finder = $this->finder();
        $this->db = $this->createMock(DB::class);
        $this->pages = new FakePages();
        $this->view = new View("./views/", XH_includeVar("./languages/en.php", "plugin_tx")["realblog"]);
    }

    private function sut(): ArticleController
    {
        return new ArticleController(
            $this->conf,
            $this->finder,
            $this->db,
            $this->pages,
            $this->view
        );
    }

    public function testRendersArticle(): void
    {
        $this->conf["show_teaser"] = "true";
        $this->pages = new FakePages(["h" => ["", "Blog"]]);
        $request = new FakeRequest([
            "url" => "http://example.com/?&function=realblog_article&realblog_id=3&realblog_search=word"
                . "&realblog_page=1&realblog_selected=Blog",
            "s" => 1,
        ]);
        $response = $this->sut()($request, "blog", true, "all");
        Approvals::verifyHtml($response->output());
    }

    public function testSetsTitleAndDescription(): void
    {
        $this->pages = new FakePages(["h" => ["", "Blog"]]);
        $request = new FakeRequest([
            "url" => "http://example.com/?&function=realblog_article&realblog_id=3",
            "s" => 1,
        ]);
        $response = $this->sut()($request, "blog", true, "all");
        $this->assertEquals("Title", $response->title());
        $this->assertEquals("Teaser", $response->description());
    }

    public function testRecordsPageView(): void
    {
        $this->db = $this->createMock(DB::class);
        $this->db->expects($this->once())->method("recordPageView")->with(3);
        $this->pages = new FakePages(["h" => ["", "Blog"]]);
        $request = new FakeRequest([
            "url" => "http://example.com/?&function=realblog_article&realblog_id=3",
            "s" => 1,
        ]);
        $this->sut()($request, "blog", false, "all");
    }

    public function testRendersArticleWithComments()
    {
        $this->conf["comments_plugin"] = "Realblog\\Infra";
        $this->pages = new FakePages(["h" => ["", "Blog"]]);
        $this->finder = $this->finder(["commentable" => true]);
        $request = new FakeRequest([
            "url" => "http://example.com/?&function=realblog_article&realblog_id=3&realblog_search=word"
                . "&realblog_page=1&realblog_selected=Blog",
            "s" => 1,
        ]);
        $response = $this->sut()($request, "blog", true, "all");
        Approvals::verifyHtml($response->output());
    }

    public function testRendersArticleWithCommentsAndAdminFeatures()
    {
        $this->conf["comments_plugin"] = "Realblog\\Infra";
        $this->conf["heading_above_meta"] = "true";
        $this->pages = new FakePages(["h" => ["", "Blog"]]);
        $this->finder = $this->finder(["commentable" => true]);
        $request = new FakeRequest([
            "url" => "http://example.com/?&function=realblog_article&realblog_id=3"
                . "&realblog_page=1&realblog_selected=Blog",
            "admin" => true,
            "s" => 1,
        ]);
        $response = $this->sut()($request, "blog", true, "all");
        Approvals::verifyHtml($response->output());
    }

    public function testRendersArchivedArticle(): void
    {
        $this->finder = $this->finder(["article" => $this->archivedArticle()]);
        $this->pages = new FakePages(["h" => ["irrelevant0", "irrelevant1", "Archive"]]);
        $request = new FakeRequest([
            "url" => "http://example.com/?&function=realblog_article&realblog_id=3"
                . "&realblog_year=2022&realblog_selected=Archive",
            "s" => 2,
        ]);
        $response = $this->sut()($request, "archive", true);
        Approvals::verifyHtml($response->output());
    }

    private function finder($options = [])
    {
        $finder = $this->createStub(Finder::class);
        $finder->method("countArticlesWithStatus")->willReturn($options["count"] ?? 0);
        $finder->method("findArticles")->willReturn($options["articles"] ?? []);
        $finder->method("findById")->willReturn($options["article"] ?? $this->article());
        $finder->method("findArchiveYears")->willReturn([2020, 2022]);
        $finder->method("findArchivedArticlesInPeriod")->willReturn($options["articles"] ?? []);
        $finder->method("findArchivedArticlesContaining")->willReturn([]);
        return $finder;
    }

    private function article(): FullArticle
    {
        return new FullArticle(
            3,
            1,
            gmmktime(12, 0, 0, 6, 23, 2023),
            gmmktime(12, 0, 0, 6, 23, 2023),
            gmmktime(12, 0, 0, 6, 23, 2023),
            1,
            ",test,",
            "Title",
            "Teaser",
            "Body",
            false,
            true
        );
    }

    private function archivedArticles(): array
    {
        $articles = [];
        foreach (range(2, 5) as $num) {
            $month = $num;
            $year = 2022;
            $articles[] = new Article(
                $num,
                gmmktime(12, 0, 0, $month, 14, $year),
                2,
                "",
                "Title $num",
                "Teaser $num",
                false,
                false,
                false
            );
        }
        return $articles;
    }

    private function archivedArticle(): FullArticle
    {
        return new FullArticle(
            3,
            2,
            gmmktime(12, 0, 0, 6, 23, 2022),
            gmmktime(12, 0, 0, 6, 23, 2022),
            gmmktime(12, 0, 0, 6, 23, 2022),
            2,
            "",
            "Title",
            "Teaser",
            "Body",
            false,
            false
        );
    }
}
