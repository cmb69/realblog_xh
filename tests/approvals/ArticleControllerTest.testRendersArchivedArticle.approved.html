<!-- realblog article -->
<div class="realblog_article">

  <div class="realblog_show_box">

    <div class="realblog_buttons">
      <a class="realblog_button" href="/?Archive&amp;realblog_year=2022">archive overview</a>
    </div>

    <h1>Title</h1>
    <div class="realblog_article_meta">
      <span class="realblog_meta_date">Published on 23.06.2022.</span>
    </div>
    <div class="realblog_show_story_entry">Body</div>

    <div class="realblog_buttons">
      <a class="realblog_button" href="/?Archive&amp;realblog_year=2022">archive overview</a>
    </div>

  </div>


</div>
